src/kvstore/kvstore_nccl.h

```diff
@@ -0,0 +1,182 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/**
+ * @file   kvstore_nccl.h
+ * @brief  NCCL implementation of KVStore
+ */
+#ifndef MXNET_KVSTORE_KVSTORE_NCCL_H_
+#define MXNET_KVSTORE_KVSTORE_NCCL_H_
+
+#if MXNET_USE_NCCL
+
+#include <mxnet/kvstore.h>
+#include <nccl.h>
+#include <unordered_map>
+#include <bitset>
+#include <vector>
+#include <string>
+#include <utility>
+#include <functional>
+#include <algorithm>
+#include "./comm.h"
+#include "./kvstore_local.h"
+
+
+namespace mxnet {
+namespace kvstore {
+
+/**
+ * \brief store data in local machine using NCCL
+ */
+class KVStoreNCCL : public KVStoreLocal {
```

Comment:
Does it mean multi-machine with GPUs cannot benefit from nccl?

Reply:
Currently no - it is a future work. The biggest problem is how to bootstrap NCCL in multi-node scenario, and I do not yet understand MXNet's distributed kvstore enough to use it for that task.